src/lib/storage/boot_requirements_strategies/legacy.rb

```diff
@@ -21,18 +21,73 @@
 # To contact SUSE LLC about this file by physical or electronic mail, you may
 # find current contact information at www.suse.com.
 
-require "yast"
 require "storage/boot_requirements_strategies/base"
-require "storage/planned_volume"
-require "storage/planned_volumes_list"
-require "storage/disk_size"
 
 module Yast
   module Storage
     module BootRequirementsStrategies
       # Strategy to calculate the boot requirements in a legacy system (x86
       # without EFI)
       class Legacy < Base
+
+        using Refinements::Disk
+
+        def needed_partitions
+          volumes = super
+          volumes << grub_volume if grub_partition_required? && grub_partition_missing?
+          if mbr_gap_required?
+            mbr_gap = disk_analyzer.mbr_gap[settings.root_device]
+            # fail if gap is too small
+            if mbr_gap < DiskSize.kiB(256)
+              raise Proposal::Error
+            end
+          end
+          volumes
+        end
+
+      protected
+
+        def grub_partition_missing?
+          partitions = disk_analyzer.grub_partitions[settings.root_device]
+          partitions.nil? || partitions.empty?
+        end
+
+        # only needed on GPT partition table
+        # FIXME: currently nobody creates a partition table on an empty disk...
+        def grub_partition_required?
```

Comment:
For me it makes sense that `DiskAnalyzer` also returns the type of partition present in each candidate disk, instead of having this check here.

Actually, in #46 I propose to index by disk most information returned by `DiskAnalyzer`